platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/DefaultStateGarbageCollector.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (C) 2024 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.platform.state.signed;
+
+import static com.swirlds.common.units.TimeUnit.UNIT_MICROSECONDS;
+import static com.swirlds.common.units.TimeUnit.UNIT_NANOSECONDS;
+import static com.swirlds.logging.legacy.LogMarker.EXCEPTION;
+
+import com.swirlds.common.context.PlatformContext;
+import com.swirlds.common.metrics.RunningAverageMetric;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * This class is responsible for the deletion of signed states. In case signed state deletion is expensive, we never
+ * want to delete a signed state on the last thread that releases it.
+ */
+public class DefaultStateGarbageCollector implements StateGarbageCollector {
+
+    private static final Logger logger = LogManager.getLogger(DefaultStateGarbageCollector.class);
+
+    private final List<SignedState> states = new LinkedList<>();
+
+    private static final RunningAverageMetric.Config UNDELETABLE_STATES_CONFIG = new RunningAverageMetric.Config(
```

Comment:
Thinking aloud - is "undeletable" the right term here? "Undeletable" somehow sounds permanent. What do you think about `StatesIneligibleForDeletion`?

No big deal either way, if you'd prefer to just leave it

Reply:
changed to suggested name